executor/prebackup.go

```diff
@@ -0,0 +1,142 @@
+package executor
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	k8upv1alpha1 "github.com/vshn/k8up/api/v1alpha1"
+)
+
+// StartPreBackup will start the defined pods as deployments.
+// It returns true and no error if there are no PreBackups to run after setting the ConditionPreBackupPodsReady accordingly.
+// Returns false and error if the retrieval of PreBackupPod templates failed (with status update).
+func (b *BackupExecutor) StartPreBackup() (bool, error) {
+
+	templates, err := b.fetchPreBackupPodTemplates()
+	if err != nil {
+		b.SetConditionFalseWithMessage(k8upv1alpha1.ConditionPreBackupPodsReady, k8upv1alpha1.ReasonRetrievalFailed, "error while retrieving container definitions: %v", err.Error())
+		return false, err
+	}
+
+	if len(templates.Items) == 0 {
+		b.SetConditionTrueWithMessage(k8upv1alpha1.ConditionPreBackupPodsReady, k8upv1alpha1.ReasonNoPreBackupPodsFound, "no container definitions found")
+		return true, nil
+	}
+
+	deployments := b.generateDeployments(templates.Items)
+
+	return b.startAllPreBackupDeployments(deployments)
+}
+
+// startAllPreBackupDeployments attempts to start all given deployment templates.
+// If some of them are already existing, their status will determine whether the backup is considered failed.
+// It returns true if all deployments are ready, also the ConditionPreBackupPodsReady will be set to ReasonReady.
+// Otherwise it returns false, indicating that not all PreBackup deployment(s) are ready.
+func (b *BackupExecutor) startAllPreBackupDeployments(deployments []appsv1.Deployment) (bool, error) {
+	existingDeployments := make([]*appsv1.Deployment, 0)
+	for _, deployment := range deployments {
+		exists, err := b.createPreBackupDeployment(deployment)
+		if err != nil {
+			return false, err
+		}
+		if exists {
+			deployment, err := b.fetchPreBackupDeployment(deployment.Name, deployment.Namespace)
+			if err != nil {
+				return false, err
+			}
+			existingDeployments = append(existingDeployments, deployment)
+		} else {
+			b.Log.Info("started pre backup pod", "namespace", deployment.Namespace, "name", deployment.Name)
+		}
```

Comment:
Maybe switch this:

```suggestion
		if !exists {
			b.Log.Info("started pre backup pod", "namespace", deployment.Namespace, "name", deployment.Name)
			continue
		}
		
		deployment, err := b.fetchPreBackupDeployment(deployment.Name, deployment.Namespace)
		if err != nil {
			return false, err
		}
		existingDeployments = append(existingDeployments, deployment)
```

Reply:
jup, that's easier to read. thanks!

Reply:
I was able to optimize even more and combine into `fetchOrCreate` function, so no additional list either. should be even more understandable now